### PR TITLE
[IIO]Tensor src iio start stop functions

### PR DESCRIPTION
**Pull request by kparichay**
1. Add reading device and its properties
2. Get device ready to data capture, start function (null->ready state)

Helper functions to read files and directories, parse strings are also added
implementing for #1199 

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

### CODE REVIEW

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 1]
```diff
@@ -260,15 +728,227 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+      goto error_trigger_free;
+    }
+  }
+  // setup the frequency (only verifying the frequency now)
+  // @todo verify setting up frequency
+  guint sampling_frequency =
+      gst_tensor_src_iio_set_frequency (self->device.base_dir,
+      self->sampling_frequency);
+  if (0 == sampling_frequency) {
+    GST_ERROR_OBJECT (self, "IIO device does not support %ld frequency\n",
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Not compatible with 32bit systems

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 2]
```diff
@@ -217,28 +282,390 @@ gst_tensor_src_iio_class_init (GstTensorSrcIIOClass * klass)
...
+/**
+ * @brief get type info about the channel from the string
+ */
+static gboolean
+gst_tensor_src_iio_set_channel_type (GstTensorSrcIIOChannelProperties * prop,
+    const gchar * file_contents)
+{
+  gchar endianchar, signchar;
+  gint arguments_filled;
+  arguments_filled = sscanf ("%ce:%c%u/%u>>%u", &endianchar, &signchar,
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: missing the "str" argument for sscanf:

```
int sscanf(const char *str, const char *format, ...);
```

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 3]
```diff
@@ -217,28 +282,390 @@ gst_tensor_src_iio_class_init (GstTensorSrcIIOClass * klass)
...
+}
+
+/**
+ * @brief get type info about the channel from the string
+ */
+static gboolean
+gst_tensor_src_iio_set_channel_type (GstTensorSrcIIOChannelProperties * prop,
+    const gchar * file_contents)
+{
+  gchar endianchar, signchar;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: initialize these two before using them although sscanf is supposed to set them.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 2 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 3 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 1 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: done

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 4]
```diff
@@ -965,8 +965,16 @@ gst_tensor_src_iio_start (GstBaseSrc * src)
...
-  /** free resources */
-  //FIXME: fill this function
+  /** free resources related to the device */
+  GstTensorSrcIIO *self;
+  self = GST_TENSOR_SRC_IIO_CAST (src);
+
+  self->configured = FALSE;
+  g_list_free_full (self->channels, gst_tensor_src_iio_channel_properties_free);
+  if (self->trigger.base_dir != NULL) {
+    g_free (self->trigger.base_dir);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: 1. I'd recommend to set it NULL after g_free.
2. You do not need NULL check for g_free. g_free does NULL check internally.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 5]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+  struct dirent *dir_entry;
+  gchar *filename = NULL;
+  guint id = -1;
+  gchar *file_contents = NULL;
+  guint ret = -1;
+
+  dptr = opendir (dir_name);
+  if (G_UNLIKELY (NULL == dptr)) {
+    GST_ERROR ("No IIO %s available.", prefix);
+    goto error_close_dir;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Can you close NULL dptr?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 6]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+gst_tensor_src_iio_get_id_by_name (const gchar * dir_name, const gchar * name,
+    const gchar * prefix)
+{
+  DIR *dptr = NULL;
+  GError *error = NULL;
+  struct dirent *dir_entry;
+  gchar *filename = NULL;
+  guint id = -1;
+  gchar *file_contents = NULL;
+  guint ret = -1;
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: -1 for unsigned int?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 7]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
   return;
 }
 
+/**
+ * @brief check if device/trigger with the given name exists
+ * @return >=0 if OK, represents device/trigger number
+ *         -1  if returned with error
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: the function return type is unsigned.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 8]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+      &prop->mask_bits, &prop->storage_bits, &prop->shift);
+  if (arguments_filled < 5) {
+    return FALSE;
+  }
+
+  if (endianchar == 'b') {
+    prop->big_endian = TRUE;
+  } else {
+    prop->big_endian = FALSE;
+  }
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: If it's neither 'b' or 'l', let's return FALSE.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 9]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+  if (signchar == 's') {
+    prop->is_signed = TRUE;
+  } else {
+    prop->is_signed = FALSE;
+  }
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: If it's neither 's' or 'u', let's return FALSE.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 10]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+  guint ret = -1;
+  guint value;
+  guint num_channels_enabled = 0;
+
+  if (!g_file_test (dir_name, G_FILE_TEST_IS_DIR)) {
+    GST_ERROR ("No channels available.");
+    return ret;
+  }
+
+  dptr = opendir (dir_name);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Let's check if it's NULL or not. (it can pass g_file_test and fail opendir. e.g., permission error)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 11]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+  GError *error = NULL;
+  gchar *filename = NULL;
+  gchar *file_contents = NULL;
+  guint i = 0;
+  gulong ret = 0, val = -1;
+
+  // get frequency list supported by the device
+  filename = g_strdup_printf ("%s/%s", base_dir, AVAIL_FREQUENCY_FILE);
+  if (!g_file_get_contents (filename, &file_contents, NULL, &error)) {
+    GST_ERROR ("Unable to read sampling frequency for device %s\n.", base_dir);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: \n. --> .\n ? :)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 12]
```diff
@@ -288,6 +331,295 @@ gst_tensor_src_iio_init (GstTensorSrcIIO * self)
...
+    g_error_free (error);
+    goto del_filename;
+  }
+
+  gchar **freq_list = g_strsplit (file_contents, " ", -1);
+  gint num = g_strv_length (freq_list);
+  if (num == 0) {
+    GST_ERROR ("No sampling frequencies for device %s\n.", base_dir);
+    g_error_free (error);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: I don't think error is available here.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 13]
```diff
@@ -383,7 +715,7 @@ gst_tensor_src_iio_get_property (GObject * object, guint prop_id,
       break;
 
     case PROP_FREQUENCY:
-      g_value_set_uint64 (value, self->sampling_frequency);
+      g_value_set_ulong (value, self->sampling_frequency);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: As Linux kernel uses unsigned-long for sampling frequency, we need to be ready for different bit-width for this property as you are currently doing.

However, please make sure that the results (srcpad) of this plugin consistent regardless of architecture. The pipeline itself may be distributed over different computers. (e.g., tensor_iio_src in 32bit system, it's user tensor_filter in 64bit system)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 14]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+ */
+static gboolean
+gst_tensor_write_sysfs_string (GstTensorSrcIIO * self, const gchar * file,
+    const gchar * base_dir, const gchar * contents)
+{
+  gchar *filename = NULL;
+  GError *error = NULL;
+  gboolean ret = FALSE;
+
+  filename = g_strdup_printf ("%s/%s", base_dir, file);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Don't join paths with this.
Use Glib.Path for portability.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 15]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+
+  if (DBG) {
+    gchar *file_contents = NULL;
+    ret = FALSE;
+    if (!g_file_get_contents (filename, &file_contents, NULL, &error)) {
+      GST_ERROR ("Unable to read file %s with error %s\n", filename,
+          error->message);
+      g_error_free (error);
+    } else {
+      if (g_strcmp0 (contents, file_contents)) {
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Missing ! ?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 16]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
   /** load and init resources */
-  //FIXME: fill this function
+  GstTensorSrcIIO *self;
+  self = GST_TENSOR_SRC_IIO_CAST (src);
+  guint id;
+
+  // Find the device
+  id = gst_tensor_src_iio_get_id_by_name (IIO_BASE_DIR, self->device.name,
+      DEVICE_PREFIX);
+  if (G_UNLIKELY (id < 0)) {
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: id is uint.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 17]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+
+  // no more configuration for one shot mode
+  if (!g_strcmp0 (self->mode, MODE_ONE_SHOT)) {
+    goto safe_return;
+  }
+  // register the trigger
+  if (self->trigger.name != NULL) {
+    // verify if trigger is supported by our device
+    gchar *trigger_device_dir =
+        g_strdup_printf ("%s%s", self->device.base_dir, TRIGGER);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Use Glib.Path for pathname operations.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 18]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+  // Find the device
+  id = gst_tensor_src_iio_get_id_by_name (IIO_BASE_DIR, self->device.name,
+      DEVICE_PREFIX);
+  if (G_UNLIKELY (id < 0)) {
+    GST_ERROR_OBJECT (self, "Cannot find the IIO device with name: %s\n",
+        self->device.name);
+    goto error_return;
+  }
+  self->device.id = id;
+  self->device.base_dir = g_strdup_printf ("%s%s%d/", IIO_BASE_DIR,
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Don't add '/' to path names in the code. It's not portable.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 13 above)
Comment (myungjoo) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: I'd recommend to use 64bit internal representation and accept both 32/64 representations for inputs from Linux IIO.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 19]
```diff
@@ -401,15 +733,230 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+          self->device.name);
+      g_free (trigger_device_dir);
+      goto error_device_free;
+    }
+    g_free (trigger_device_dir);
+
+    // find if the provided trigger exists
+    id = gst_tensor_src_iio_get_id_by_name (IIO_BASE_DIR, self->trigger.name,
+        TRIGGER_PREFIX);
+    if (G_UNLIKELY (id < 0)) {
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: again. as id is unsigned, this becomes inactive.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 4 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 5 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: corrected.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 6 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Corrected here as well as at other places.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 7 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Corrected here as well as at other places.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 9 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: done

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 8 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: done

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 11 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: done.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 12 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: yes, corrected.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 14 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated accordingly.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 15 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Yes, corrected.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 16 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Corrected here as well as at other places.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 17 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated here as well as at other places.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 18 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Yes
updated accordingly.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 19 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 10 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 13 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Updated.
Internally treated with uint64. Exposure to outside access is constrained with ulong.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 20]
```diff
@@ -260,15 +757,260 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+static gboolean
+gst_tensor_write_sysfs_string (GstTensorSrcIIO * self, const gchar * file,
+    const gchar * base_dir, const gchar * contents)
+{
+  gchar *filename = NULL;
+  GError *error = NULL;
+  gboolean ret = FALSE;
+
+  filename = g_build_filename (base_dir, file, NULL);
+  if (!g_file_set_contents (filename, contents, strlen (contents), &error)) {
```
Comment (kparichay) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Can g_file_set_contents() create problems here?
g_file_set_contents() recreates the file, removing existing file related metadata.

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: using fprintf() instead.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 21]
```diff
@@ -260,15 +757,274 @@ gst_tensor_src_iio_finalize (GObject * object)
...
+  filename = g_build_filename (base_dir, file, NULL);
+  fd = fopen (filename, "w");
+  if (fd == NULL) {
+    GST_ERROR ("Unable to open file to write %s.\n", filename);
+    goto error_free_filename;
+  }
+
+  bytes_printed = fprintf (fd, "%s", contents);
+  if (bytes_printed != strlen (contents)) {
+    GST_ERROR ("Unable to write to file %s.\n", filename);
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Suggestion:
Later, please consider GST_ERROR_OBJECT (self, .. ) for filtering logs with element name.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 22]
```diff
@@ -260,15 +757,274 @@ gst_tensor_src_iio_finalize (GObject * object)
...
   /** load and init resources */
-  //FIXME: fill this function
+  GstTensorSrcIIO *self;
+  self = GST_TENSOR_SRC_IIO_CAST (src);
+  gint id;
+  guint64 sampling_frequency;
+  gchar *dirname = NULL;
+  gchar *filename = NULL;
+
+  // Find the device
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Now meson default option 'gnu89'  does not make an error for line comment with  //.
Please change all to /* */ in next PR.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 21 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Noted, will update in next PR.
I think would be not very clean with the helper functions as using GST_ERROR_OBJECT (self, ..) would require passing self object to all the helper functions.
Is there any better way to handle this?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 22 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Noted, will update in next PR.

**Comment by kparichay** [comment]
> LGTM. Please rebase. (squash if the last commit does not seem to be cleanly generated)

Done.

**Comment by myungjoo** [comment]
> [IIO] Add properties for IIO kparichay committed 2 days ago
[IIO] Add start/stop for src IIO kparichay committed 2 days ago
[IIO] Add start/stop for src IIO kparichay committed 2 days ago
[IIO] Added doxygen entries kparichay committed a day ago

You need to clean up and rearrange commits in this PR. (git rebase -i HEAD~5 and try squash a few of them)

And, you need to drop the first commit of this series; it has unneeded and undesirable code line.

**Comment by kparichay** [comment]
> You need to clean up and rearrange commits in this PR. (git rebase -i HEAD~5 and try squash a few of them)
> 
> And, you need to drop the first commit of this series; it has unneeded and undesirable code line.

Updated.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 23]
```diff
@@ -52,6 +52,7 @@
 #include <gst/gst.h>
 #include <glib.h>
 #include <string.h>
+#include <stdio.h>
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: Please do not add unrelated changes.
The topic of this commit is ```[IIO] Added doxygen entries ```

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: my mistake
updated with changes applied to appropriate commit.
